merkle-tree/merkle_tree.go

```diff
@@ -0,0 +1,225 @@
+	// rootNode is the root node of the tree
+	rootNode *MerkleNode
+	// leaf nodes is the list of leaf nodes of the tree (the lowest level of the tree)
+	leafNodes []*MerkleNode
+}
+
+// NewMerkleTree creates a new Merkle tree from the provided data and using the default hashing (Keccak256).
+func NewMerkleTree(data [][]byte) (*MerkleTree, error) {
+	return NewMerkleTreeWithHashing(data, crypto.NewKeccakState())
+}
+
+// NewMerkleTreeWithHashing creates a new Merkle tree from the provided data and hash type
+func NewMerkleTreeWithHashing(data [][]byte, hasher hash.Hash) (*MerkleTree, error) {
+	if len(data) == 0 {
+		return nil, errors.New("tree must contain at least one leaf")
+	}
+
+	leafNodes := make([]*MerkleNode, len(data))
+	for i, d := range data {
+		leafNodes[i] = newMerkleNode(nil, nil, d, hasher)
+	}
+
+	nodes := leafNodes
+	for len(nodes) > 1 {
+		var newLevel []*MerkleNode
+
+		for i := 0; i < len(nodes); i += 2 {
+			left := nodes[i]
+			right := nodes[i] // if the tree has uneven number of nodes, this will duplicate the last one
+
+			if i+1 < len(nodes) {
+				right = nodes[i+1] // if it has even numbers, then this will point to the last node
+			}
+
+			parent := newMerkleNode(left, right, nil, hasher)
+			left.parent = parent
+			right.parent = parent
+
+			newLevel = append(newLevel, parent)
+		}
+
+		nodes = newLevel
+	}
+
+	return &MerkleTree{
+		hasher:    hasher,
+		rootNode:  nodes[0],
+		leafNodes: leafNodes,
+	}, nil
+}
+
+// LeafIndex returns the index of given leaf if found in tree
+func (t *MerkleTree) LeafIndex(leaf []byte) (uint64, error) {
+	for i, leafNode := range t.leafNodes {
+		if bytes.Equal(leafNode.data, leaf) {
+			return uint64(i), nil
+		}
+	}
+
+	return 0, errLeafNotFound
+}
+
+// Hash is the Merkle Tree root hash
+func (t *MerkleTree) Hash() types.Hash {
+	return types.BytesToHash(t.rootNode.hash)
+}
+
+// String implements the stringer interface
+func (t *MerkleTree) String() string {
+	return hex.EncodeToString(t.Hash().Bytes())
+}
+
+// Depth returns the depth of merkle tree
+func (t *MerkleTree) Depth() int {
+	return int(math.Ceil(math.Log2(float64(len(t.leafNodes)))))
+}
+
+// GenerateProof generates the proof of membership for a piece of data in the Merkle tree.
+func (t *MerkleTree) GenerateProof(leaf []byte) ([]types.Hash, error) {
+	proof := []types.Hash{}
+
+	leafNode := t.findLeafNode(leaf)
+	if leafNode == nil {
+		return nil, fmt.Errorf("given data not in merkle tree")
+	}
+
+	node := leafNode
+	for !bytes.Equal(node.hash, t.rootNode.hash) {
+		if bytes.Equal(node.parent.left.hash, node.hash) {
+			proof = append(proof, types.BytesToHash(node.parent.right.hash))
+		} else {
+			proof = append(proof, types.BytesToHash(node.parent.left.hash))
+		}
+
+		node = node.parent
+	}
+
+	return proof, nil
+}
+
+// findLeafNode finds the given leaf node that corresponds to given leaf data
+func (t *MerkleTree) findLeafNode(leaf []byte) *MerkleNode {
+	for _, leafNode := range t.leafNodes {
```

Comment:
Does it make sense introducing a `map[types.Hash]*MerkleNode` and rely on it when searching for `MerkleNode` instead of looping through all leaf nodes. WDYT?

So when adding `MerkleNode` to the map, we would just hash data using provided hasher and put it into the map. And here, when looking up for a node, we would again hash provided leaf data and just access to the map and return a `MerkleNode` if found.

Reply:
> theoretically, you can have leaf nodes that contain the same data (meaning their hashes will be the same), so the map won't work in that case.

If leaf data are going to be the same, then the outcome is the same. But it is faster to look up in the map for large amount of leaf nodes.

Reply:
Ok, I see your point, My reasoning is that it if we use a map, then `GetLeafIndex` function will not work. If we are to keep leafs in both slice and map, then there should be a reason for it (generating proof for large trees is to slow, for example). I'll write some benchmark tests, and we will see if generating proof is slow in large trees. If it is, using a map makes sense, if it's reasonable, then I would say lets leave it as it is.

Reply:
I've pushed the Benchmark tests, since they are a nice thing to have. I've added the map as you suggested and didn't see any improvements, so let's leave it as is.